### Sprint 9 Project

* Note to reviewer:
My sincerest apologies, I did not delete your comments at all.  I had trouble figuring out how to download your revised version of my project with corrections, so I just examined your comments and made new corrections on my existing notebook.  My apologies for the inconvenience.

### Summary:
This project contains information for Regions 0, 1, and 2 for the OilyGiant mining company.  In this project, I will be using a Linear Regression model in order to find which of the 3 regions listed is the most profitable for the OilyGiant mining company.  This project involves calculating the estimated profit for each region using a Linear Regression Model and then using the bootstrapping to revaluate the average estimed profit once again from a dataset of 1000 bootstrapped samples with replacement.

Corrections:
X Correct conclusions and code results after dropping duplicate values. *done*
* Create charts for distribution of dimensions for each region.
X Compare predicted reserves with break even volume. *done*
X Fix bootstrapping codes (make it all one code). *done* 
* Fix final conclusion
* Make new captions for duplicate values.

### Step 1: Download and Prepare the Data. Explain the Procedure

In [190]:
#Import the necessary libraries.
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

from sklearn.model_selection import RandomizedSearchCV

from sklearn.pipeline import Pipeline

from sklearn.preprocessing import StandardScaler

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

from sklearn.model_selection import cross_val_score

import numpy as np

In [191]:
#Use the pd.read_csv() method in order to create a dataframe for Region 0.
try:
    r0 = pd.read_csv('geo_data_0.csv')
except:
    r0 = pd.read_csv('/datasets/geo_data_0.csv')
     

In [192]:
#Use the head() method to check the first five lines of the r0 dataframe.
r0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

Code Result:
It appears that the r0 dataframe has been sucessfully uploaded.

In [193]:
#Use the pd.read_csv() method in order to create a dataframe for Region 1.
try: 
    r1 = pd.read_csv('geo_data_1.csv')
except:
    r1 = pd.read_csv('/datasets/geo_data_1.csv')

In [194]:
#Use the head() method to check the first five lines of the r1 dataframe.
r1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

Code Result:
It appears that the r1 dataframe has been sucessfully uploaded.

In [195]:
#Use the pd.read_csv() method in order to create a dataframe for Region 2.
try:
    r2 = pd.read_csv('geo_data_2.csv')
except:
    r2 = pd.read_csv('/datasets/geo_data_2.csv')

In [196]:
#Use the head() method to check the first five lines of the r2 dataframe.
r2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Code Result:
It appears that the r2 dataframe has been sucessfully uploaded.

#### Data Checking

In [197]:
#Use the isna() and sum() methods to count number of missing values in the r0 dataframe.
r0.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Code Result:
There are no missing values noted in the r0 dataframe.

In [198]:
#Use the isna() and sum() methods to count number of missing values in the r1 dataframe.
r1.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Code Result:
There are no missing values noted in the r2 dataframe.

In [199]:
#Use the isna() and sum() methods to count number of missing values in the r2 dataframe.
r2.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

#Use the isna() and sum() methods to count number of missing values in the r2 dataframe.

In [200]:
r0['id'].duplicated().sum()

10

In [201]:
r0 = r0.drop_duplicates(subset='id')

r0['id'].duplicated().sum()

0

In [202]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r0 dataframe.
r0.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r0 dataframe.

In [203]:
r1['id'].duplicated().sum()

4

In [204]:
r1 = r1.drop_duplicates(subset='id')

r1['id'].duplicated().sum()

0

In [205]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r1 dataframe.
r1.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r1 dataframe.

In [206]:
r2['id'].duplicated().sum()

4

In [207]:
r2 = r2.drop_duplicates(subset='id')

r2['id'].duplicated().sum()

0

In [208]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r2 dataframe.
r2.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r2 dataframe.

In [209]:
#Use the info() method in order check the data types in the r0 dataframe.
r0.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99990 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99990 non-null  object 
 1   f0       99990 non-null  float64
 2   f1       99990 non-null  float64
 3   f2       99990 non-null  float64
 4   product  99990 non-null  float64
dtypes: float64(4), object(1)
memory usage: 4.6+ MB


Code Result:
The data types in the r0 dataframe appear to be listed correctly.

In [210]:
#Use the info() method in order check the data types in the r1 dataframe.
r1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99996 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99996 non-null  object 
 1   f0       99996 non-null  float64
 2   f1       99996 non-null  float64
 3   f2       99996 non-null  float64
 4   product  99996 non-null  float64
dtypes: float64(4), object(1)
memory usage: 4.6+ MB


Code Result:
The data types in the r1 dataframe appear to be listed correctly.

In [211]:
#Use the info() method in order check the data types in the r2 dataframe.
r2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99996 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99996 non-null  object 
 1   f0       99996 non-null  float64
 2   f1       99996 non-null  float64
 3   f2       99996 non-null  float64
 4   product  99996 non-null  float64
dtypes: float64(4), object(1)
memory usage: 4.6+ MB


Code Result:
The data types in the r2 dataframe appear to be listed correctly.

### Step 2. Train and test the model for each region.

####   2.1. Split the data into a training set and validation set at a ratio of 75:25.

In [212]:
#Prepare the features and target values with the data from the r0 dataframe.
features_r0 = r0.drop(['id', 'product'], axis=1)

target_r0 = r0['product']

In [213]:
#Use the train_test_split() method in order to split the features_r0 and target_r0 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r0, features_valid_r0, target_train_r0, target_valid_r0 = train_test_split(features_r0, target_r0, test_size=0.25, random_state=12345)


In [214]:
#Prepare the features and target values with the data from the r1 dataframe.
features_r1 = r1.drop(['id', 'product'], axis=1)

target_r1 = r1['product']

In [215]:
#Use the train_test_split() method in order to split the features_r1 and target_r1 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r1, features_valid_r1, target_train_r1, target_valid_r1 = train_test_split(features_r1, target_r1, test_size=0.25, random_state=12345)

In [216]:
#Prepare the features and target values with the data from the r2 dataframe.
features_r2 = r2.drop(['id', 'product'], axis=1)

target_r2 = r2['product']

In [217]:
#Use the train_test_split() method in order to split the features_r2 and target_r2 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r2, features_valid_r2, target_train_r2, target_valid_r2 = train_test_split(features_r2, target_r2, test_size=0.25, random_state=12345)


 #### 2.2. Train the model and make predictions for the validation set. 
 
 2.3. Save the predictions and correct answers for the validation set.

 2.4. Print the average volume of predicted reserves and model RMSE.


#### Region 0 Model

In [218]:
#Create a Linear Regression Model for Region 0 using Scaler() and LinearRegression().
r0_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [219]:
#Create a parameter grid for the LinearRegression() model.
param_grid = {'scaler__with_mean': [True, False], 'scaler__with_std': [True, False]}

In [220]:
#Use RandomizedSearchCV() in order to create a Randomized Search model in order to tune the hyperparameters for the 
#LinearRegression() model.
r0_lr_rsearch = RandomizedSearchCV(estimator=r0_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [221]:
#Use the fit() method in order to train the RandomizedSearchCV() model with the features and target from the training set for
#Region 0.
r0_lr_rsearch.fit(features_train_r0, target_train_r0)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [222]:
#Use the best_params_ and best_score_methods_ in order to print out the best score and the best hyperparameters for the Linear Regression model.
r0_lr_bparams = r0_lr_rsearch.best_params_

r0_lr_bscore = r0_lr_rsearch.best_score_

print("Region 0 Linear Regression Best Hyperparameters:", r0_lr_bparams)

print('Region 0 Linear Regression Best Score:', r0_lr_bscore)

Region 0 Linear Regression Best Hyperparameters: {'scaler__with_std': False, 'scaler__with_mean': True}
Region 0 Linear Regression Best Score: 0.27659260309400285


Code Result:
It appears that the best hyperparameters for the Linear Regression Model for Region 0 are:  {'scaler__with_std': False, 'scaler__with_mean': True} and the best score for the model is approximately 0.27.

In [223]:
#Use Pipeline() and LinearRegression() once again in order to recreate r0_model with the new tuned hyperparameters.
r0_model = Pipeline([('scaler', StandardScaler(with_std=False, with_mean=True,)), ('regressor', LinearRegression())])

In [224]:
#Use the cross_val_score() method in order to find the RMSE for r0_model.
r0_scores = cross_val_score(r0_model, features_train_r0, target_train_r0, cv=5, scoring='neg_mean_squared_error')

r0_rmse_scores = (-r0_scores) ** 0.5

r0_rmse = r0_rmse_scores.mean()

print("Linear Regression Model RMSE:", r0_rmse)

Linear Regression Model RMSE: 37.64115697063362


Code Result:
The RMSE for r0_model is: approximately 37.64.

In [225]:
#Use the fit() method in order to train r0_model with the features_train_r0 and target_train_r0 variables.
r0_model.fit(features_train_r0, target_train_r0)

Pipeline(steps=[('scaler', StandardScaler(with_std=False)),
                ('regressor', LinearRegression())])

In [226]:
#Use the predict() method on r0_model in order to make a prediction using the features_valid_r0 variable.
r0_predict = r0_model.predict(features_valid_r0)

In [227]:
#Use the mean() method on the r0_predict variable in order to calcualte the average predicted reserves for Region 0.
r0_apr = r0_predict.mean()

#Print the amount of average predicted reserves for Region 0.
print("Region 0 Average Predicted Reserves:", r0_apr)

#Print the rmse for the r0_model.
print("Linear Regression Region 0 Model RMSE", r0_rmse)

Region 0 Average Predicted Reserves: 92.78915638280621
Linear Regression Region 0 Model RMSE 37.64115697063362


Code Result:
It appears that according to the model, the average amount predicted of reserves for Region 0 is approximately 92.78 K barrels.
The RMSE for the Linear Regression Model for Region 0 (r0_model) is: approximately 37.64.


#### Region 1 Model

In [228]:
#Create a Linear Regression Model for Region 1 using Scaler() and LinearRegression().
r1_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [229]:
#Use RandomizedSearchCV() in order to create a Randomized Search model in order to tune the hyperparameters for the 
#LinearRegression() model.
r1_lr_rsearch = RandomizedSearchCV(estimator=r1_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [230]:
#Use the fit() method in order to train the RandomizedSearchCV() model with the features and target from the training set for
#Region 1.
r1_lr_rsearch.fit(features_valid_r1, target_valid_r1)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [266]:
#Use the best_params_ and best_score_methods_ in order to print out the best score and the best hyperparameters for the Linear Regression model.
r1lr_bparams = r1_lr_rsearch.best_params_
r1_lr_bscore = r1_lr_rsearch.best_score_

print("Linear Regression Region 1 Best Hyperparameters:", r1lr_bparams)

print("Linear Regression Region 1 Best Score:", r1_lr_bscore)

Linear Regression Region 1 Best Hyperparameters: {'scaler__with_std': True, 'scaler__with_mean': True}
Linear Regression Region 1 Best Score: 0.9996221090160866


Code Result:
It appears that the best hyperparameters for the Linear Regression Model for Region 1 are:  {'scaler__with_std': True, 'scaler__with_mean': True} and the best score for the model is approximately 0.99.


In [232]:
#Use Pipeline() and LinearRegression() once again in order to recreate r1_model with the new tuned hyperparameters.
r1_model = Pipeline([('scaler', StandardScaler(with_std=True, with_mean=True,)), ('regressor', LinearRegression())])

In [233]:
#Use the cross_val_score() method in order to find the RMSE for r1_model.
r1_scores = cross_val_score(r1_model, features_train_r1, target_train_r1, cv=5, scoring='neg_mean_squared_error')

r1_rmse_scores = (-r1_scores) ** 0.5

r1_rmse = r1_rmse_scores.mean()

print("Linear Regression Region 1 Model RMSE:", r1_rmse)

Linear Regression Region 1 Model RMSE: 0.8898652962423409


Code Result:
The RMSE for r1_model is: approximately 0.88.

In [234]:
#Use the fit() method in order to train r1_model with the features_train_r1 and target_train_r1 variables.
r1_model.fit(features_train_r1, target_train_r1)

Pipeline(steps=[('scaler', StandardScaler()),
                ('regressor', LinearRegression())])

In [235]:
#Use the predict() method on r1_model in order to make a prediction using the features_valid_r1 variable.
r1_predict = r1_model.predict(features_valid_r1)

In [236]:
#Use the mean() method on the r1_predict variable in order to calcualte the average predicted reserves for Region 1.
r1_apr = r1_predict.mean()

#Print the amount of average predicted reserves for Region 1.
print("Region 1 Average Predicted Reserves:", r1_apr)

#Print the rmse for the r1_model.
print("Linear Regression Region 1 Model RMSE:", r1_rmse)

Region 1 Average Predicted Reserves: 69.1783195703043
Linear Regression Region 1 Model RMSE: 0.8898652962423409


Code Result:
It appears that according to the model, the average amount predicted of reserves for Region 1 is approximately 69.17 K barrels.
The RMSE for the Linear Regression Model for Region 1 (r1_model) is: approximately 0.88.

#### Region 2 Model

In [237]:
#Create a Linear Regression Model for Region 2 using Scaler() and LinearRegression().
r2_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [238]:
#Use RandomizedSearchCV() in order to create a Randomized Search model in order to tune the hyperparameters for the 
#LinearRegression() model.
r2_lr_rsearch = RandomizedSearchCV(estimator=r2_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [239]:
#Use the fit() method in order to train the RandomizedSearchCV() model with the features and target from the training set for
#Region 2.
r2_lr_rsearch.fit(features_train_r2, target_train_r2)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [240]:
#Use the best_params_ and best_score_methods_ in order to print out the best score and the best hyperparameters for the Linear Regression model.
r2_bparams = r2_lr_rsearch.best_params_
r2_bscore = r2_lr_rsearch.best_score_

print("Linear Regression Region 2 Model Best Hyperparameters:", r2_bparams)
print("Linear Regression Model Region 2 Best Score:", r2_bscore)

Linear Regression Region 2 Model Best Hyperparameters: {'scaler__with_std': True, 'scaler__with_mean': True}
Linear Regression Model Region 2 Best Score: 0.19970983177482343


Code Result:
It appears that the best hyperparameters for the Linear Regression Model for Region 2 are:  {'scaler__with_std': True, 'scaler__with_mean': True} and the best score for the model is approximately 0.19.

In [241]:
#Use Pipeline() and LinearRegression() once again in order to recreate r2_model with the new tuned hyperparameters.
r2_model = Pipeline([('scaler', StandardScaler(with_std=True, with_mean=True,)), ('regressor', LinearRegression())])

In [242]:
#Use the cross_val_score() method in order to find the RMSE for r2_model.
r2_scores = cross_val_score(r2_model, features_train_r2, target_train_r2, cv=5, scoring='neg_mean_squared_error')

r2_rmse_scores = (-r2_scores) ** 0.5

r2_rmse = r2_rmse_scores.mean()

print("Linear Regression Region 3 Model RMSE:", r2_rmse)

Linear Regression Region 3 Model RMSE: 40.050653750165914


Code Result:
The RMSE for r2_model is: approximately 40.05.

In [243]:
#Use the fit() method in order to train r2_model with the features_train_r2 and target_train_r2 variables.
r2_model.fit(features_train_r2, target_train_r2)

Pipeline(steps=[('scaler', StandardScaler()),
                ('regressor', LinearRegression())])

In [244]:
#Use the predict() method on r2_model in order to make a prediction using the features_valid_r2 variable.
r2_predict = r2_model.predict(features_valid_r2)

In [245]:
#Use the mean() method on the r2_predict variable in order to calcualte the average predicted reserves for Region 2.
r2_apr = r2_predict.mean()

#Print the amount of average predicted reserves for Region 2.
print("Region 2 Average Predicted Reserves:", r2_apr)

#Print the rmse for the r2_model.
print("Linear Regression Model Region 2 RMSE:", r2_rmse)

Region 2 Average Predicted Reserves: 94.86572480562035
Linear Regression Model Region 2 RMSE: 40.050653750165914


Code Result:
It appears that according to the model, the average amount predicted of reserves for Region 2 is approximately 94.86 K barrels.
The RMSE for the Linear Regression Model for Region 2 (r2_model) is: approximately 40.05.


 #### 2.5. Analyze the results.

According to the results, Region 2 has the highest average amount of predicted reserves with an average predicted amount of: approximately 94.86 K  barrels and a Linear Regresion model RMSE of: approximately 40.05.  Region 0 has the 2nd highest average amount of predicted reserves with an average predicted amount of: 92.78 K barrels and a Linear Regression Model RMSE of: approximately 37.64.  Region 1 has the lowest average amount of predicted reserves with an average predicted akount of: 69.17 K barrels and a Linear Regression model RMSE of: approximately 0.88. 

### Step 3. Prepare for profit calculation:

3.1 Store All the Key Values in Separate Variables

In [246]:
#Create a variable for the budget.
budget = 100000000

#Create a variable for the amount of revenue generated per barrel.
revenue_per_barrel = 4500

#Calculate the cost per well by dividing the budget by 200.
cost_per_well = budget/200

print(f"Cost per well: {cost_per_well:.2f}")

Cost per well: 500000.00


Code Result:
The cost per well is: $50,000.

#### 3.2 Calculate the volume of reserves sufficient for developing a new well without losses

In [247]:
#Calculate the suffienct volume of reserves for developing a new well without losses by dividing the cost_per_well by 
#revenue per barrel.
sufficient_volume = cost_per_well/revenue_per_barrel

print(f"Sufficient volume of reserves for developing a new well without losses: \n{sufficient_volume:.2f}")

Sufficient volume of reserves for developing a new well without losses: 
111.11


Code Result:
It appears that the volume of reserves sufficient enough for developing a new well without losses is:
approximately 111.11 K barrels.

 #### 3.3. Provide the findings about the preparation for profit calculation step.

It appears that the cost per well is: $50,000 and the sufficient volume of reserves for developing a new well without losses is:
approximately 111.11 K barrels.

Region 0 Analysis:
It appears that Region 0 has 92.78 K barrels of reserves, which is 18.32 K less reserves than the calculated break even volume of reserves which is 111.11 K barrels.

Region 1 Analysis:
It appears that Region 1 has 69.17 K barrels of reserves, which is 41.94 K less reserves than the calculated break even volume of reserves which is 111.11 K barrels.

Region 2 Analysis:
It appears that Region 2 has 94.86 K barrels of reserves, which is 16.25 K less reserves than the calculated break even volume of reserves which is 111.11 K barrels.

Conclusion:
According to this analysis, it appears that Region 2 would be the best choice when it comes to comparing average predicted reserves to the calculated break even volume of suffient reserves needed for developing a new well without losses.  Region 2 had the least difference when subtracting average predicted reserves from the suffiencnt calculated break even volume of reserves needed for developing a new well without losses.

### Step 4. Write a function to calculate profit from a set of selected oil wells and model predictions:

 4.1. Pick the wells with the highest values of predictions. 

 4.2. Summarize the target volume of reserves in accordance with these predictions

In [248]:
#Create a dataframe for region 0 that incorporates the new predicted results and the target valid in two
#separate columns.
r0_valid_df = features_valid_r0.copy()
r0_valid_df['actual_valid'] = target_valid_r0
r0_valid_df['predicted'] = r0_predict

r0_valid_df.head()

f0        f1        f2  actual_valid   predicted
49468  0.073644  1.022011  6.064581    123.566299  103.950372
55496  1.986176  0.274681 -1.152339     63.442039   73.687869
11690  2.018384 -0.137280  2.085813     88.718000  101.015288
13770  1.884278  0.335650  2.102482     51.422060   94.048415
72395  1.102171 -0.514353 -0.209767    102.569999   87.559552

In [249]:
#Create a dataframe for region 1 that incorporates the new predicted results and the target valid in two
#separate columns.
r1_valid_df = features_valid_r1.copy()
r1_valid_df['acutal_valid'] = target_valid_r1
r1_valid_df['predicted'] = r1_predict

r1_valid_df.head()

f0        f1        f2  acutal_valid   predicted
71748  15.491360 -4.746353  1.999690     53.906522   53.407255
80490  -8.840548 -6.894209  5.000913    137.945408  137.863458
2655   -6.507568 -4.817448  1.003449     30.132364   29.745003
71171  -3.001247 -3.916052  3.006016     84.038886   83.188229
91141   6.090476 -4.494723  0.013815      0.000000    1.242101

In [250]:
#Create a dataframe for region 2 that incorporates the new predicted results and the target valid in two
#separate columns.
r2_valid_df = features_valid_r2.copy()
r2_valid_df['actual_valid'] = target_valid_r2
r2_valid_df['predicted'] = r2_predict

r2_valid_df.head()

f0        f1        f2  actual_valid   predicted
71750  1.953202 -1.826540 -0.757681     97.651721   76.457565
80492  0.875958  1.023876  4.992700     66.640660  109.417094
2655  -4.587649 -0.413199  1.649268     57.776581   89.922620
71173  1.446205  0.096865  0.522497     35.055066   83.701818
91140  0.058073  0.622573  1.070673     82.047594   86.761173

#### Revenue Function

In [251]:
#Create a function that calculates the profit from the top 200 oil wells according to the model predictions.
def calc_profit(valid_df, n_top_wells=200, barrel_price=4500, budget=100e6):
    predicted_sorted = valid_df.sort_values(by='predicted', ascending=False)
    
    top_well = predicted_sorted.head(n_top_wells)
    
    total_revenue = top_well['predicted'].sum() * barrel_price
    
    profit = total_revenue - budget
    
    return profit

In [252]:
#Use the calc_profit() funciton to calculate the profit for Region 0.
r0_profit = calc_profit(r0_valid_df)

print(f"Region 0 Estimated Profit: ${r0_profit:.2f}") 

Region 0 Estimated Profit: $39816309.11


In [253]:
#Calculate the volume of reserves for Region 0 by dividing Region 0's predicted profit by the revenue generated per barrel.
r0_reserves = r0_profit / revenue_per_barrel

print(f"Region 0 Estimated Volume of Reserves According to Predicitons: {r0_reserves:.2f}")

Region 0 Estimated Volume of Reserves According to Predicitons: 8848.07


Code Result:
It appears that the estimated volume of reserves for Region 0 is: approximately 8,848.11 K barrels.

In [254]:
#Use the calc_profit() funciton to calculate the profit for Region 1.
r1_profit = calc_profit(r1_valid_df)

print(f"Region 1 Estimated Profit: ${r1_profit:.2f}")

Region 1 Estimated Profit: $24860607.85


In [255]:
#Calculate the volume of reserves for Region 1 by dividing Region 0's predicted profit by the revenue generated per barrel.
r1_reserves = r1_profit / revenue_per_barrel

print(f"Region 1 Estimated Volume of Reserves According to Predictions: {r1_reserves:.2f}")

Region 1 Estimated Volume of Reserves According to Predictions: 5524.58


Code Result:
It appears that the estimated volume of reserves for Region 1 is: approximately 5524.58 K barrels.

In [256]:
r2_profit = calc_profit(r2_valid_df)

print(f"Region 2 Estimated Profit:${r2_profit:.2f}")

Region 2 Estimated Profit:$33629149.08


In [257]:
r2_reserves = r2_profit / revenue_per_barrel 

print(f"Region 2 Estimated Volume of Reserves According to Predicitons: {r2_reserves:.2f}")

Region 2 Estimated Volume of Reserves According to Predicitons: 7473.14


Code Result:
It appears that the estimated volume of reserves for Region 2 is: approximately 7473.14 K barrels.

 #### 4.3. Provide findings: suggest a region for oil wells' development and justify the choice. Calculate the profit for the obtained volume of reserves.
 
 According to the results, Region 0 has the highest estimated profit at: $39,816,309.11, which would make region 0 the best region for oil wells development.  Region 0 also the highest estimated volume of reserves at: approximately 8,848.17K barrels. Region 2 had the second highest estimated profit at: $33,629,149.08 and an estimated volume of reserves of: approximately 7,473.14 K barrels.  Region 1 has the lowest estimated profit at: $24,860,607.85 and an estimated volume of reserves of: approximately 5,524.58 K barrels.

### Step 5. Calculate risks and profit for each region:

     5.1. Use the bootstrapping technique with 1000 samples to find the distribution of profit.

     5.2. Find average profit, 95% confidence interval and risk of losses. Loss is negative profit, calculate it as a probability and then express as a percentage.

#### Bootstrapping

In [258]:
#Create a random state for the boostrapping function.
state = np.random.RandomState(12345)

#Create a function for generating 1000 bootstrapped samples.
def bootstrap(valid_df):
    profits = []
    
    for i in range(1000):
        subsample = valid_df.sample(n=500, replace=True, random_state=state)
        
        newsample = calc_profit(subsample)
        profits.append(newsample)
    
    profits = pd.Series(profits)
    
    return profits

In [271]:
def bootstrap_values(bootstrapped_code):
    average = bootstrapped_code.mean()
    lower = np.percentile(bootstrapped_code, [2.5])
    upper = np.percentile(bootstrapped_code, [97.5])
    risk_of_loss = np.mean((np.array(bootstrapped_code) < 0)) * 100
    
    print(f"Bootstrap Average Profit: ${average:.2f}")
    print(f"Bootstrap 95% Confidence Interval Lower Quantile: {lower}")
    print(f"Bootstrap 95% Confidence Interval Upper Quantile: {upper}")
    print(f"Bootstrap Risk of Loss: {risk_of_loss}")

#### Region 0 Bootstrapping

In [260]:
#Use bootstrap() to bootstrap the data in r0_valid_df.
bootstrap_0 = bootstrap(r0_valid_df)

In [272]:
region_0_values = bootstrap_values(bootstrap_0)

region_0_values

Bootstrap Average Profit: $3815855.47
Bootstrap 95% Confidence Interval Lower Quantile: [1597195.82188425]
Bootstrap 95% Confidence Interval Upper Quantile: [6084254.06319825]
Bootstrap Risk of Loss: 0.1


#### Region 1 Bootstrapping

In [262]:
#Use bootstrap() to bootstrap the data in r1_valid_df.
bootstrap_1 = bootstrap(r1_valid_df)

In [273]:
region_1_values = bootstrap_values(bootstrap_1)

region_1_values

Bootstrap Average Profit: $4762538.36
Bootstrap 95% Confidence Interval Lower Quantile: [757716.90174588]
Bootstrap 95% Confidence Interval Upper Quantile: [9120997.5742184]
Bootstrap Risk of Loss: 0.8


#### Region 2 Bootstrapping

In [264]:
#Use bootstrap() to bootstrap the data in r2_valid_df.
bootstrap_2 = bootstrap(r2_valid_df)

In [274]:
region_2_values = bootstrap_values(bootstrap_2)

region_2_values

Bootstrap Average Profit: $2916506.27
Bootstrap 95% Confidence Interval Lower Quantile: [915557.80585372]
Bootstrap 95% Confidence Interval Upper Quantile: [4918087.03264175]
Bootstrap Risk of Loss: 0.1


Region 0 Bootstrap Results:
Bootstrap Average Profit: $3,815,855.47
Bootstrap 95% Confidence Interval Lower Quantile: [1597195.82188425]
Bootstrap 95% Confidence Interval Upper Quantile: [6084254.06319825]
Bootstrap Risk of Loss: 0.1

Region 1 Bootstrap Results:
Bootstrap Average Profit: $4,762,538.36
Bootstrap 95% Confidence Interval Lower Quantile: [757716.90174588]
Bootstrap 95% Confidence Interval Upper Quantile: [9120997.5742184]
Bootstrap Risk of Loss: 0.8

Region 2 Bootstrap Results:
Bootstrap Average Profit: $2,916,506.27
Bootstrap 95% Confidence Interval Lower Quantile: [915557.80585372]
Bootstrap 95% Confidence Interval Upper Quantile: [4918087.03264175]
Bootstrap Risk of Loss: 0.1

#### 5.3. Provide findings: suggest a region for development of oil wells and justify the choice.

According to the boostrapped data, although the Region 0 is the best choice for a deveplopment of oil wells because it has the highest bootstrap average profit of: $3,815,855.47, a low risk of loss at 0.1%, and a mean confidence interval of: approximately 
$1,597,195-82-$6,084,254.06.  Region 2 would be the second best choice for development of oil wells because it has an average bootstrap profit of: $2,916,506.27, a low risk of loss of 0.1% and a mean confidence interval of: approximately $915,557.80-$4,918,087.03  .  Region 1 would be the least desirable choice for development of oil wells because it has a bootstrap average profit of: 
$4,762,538.36 (which is lower than both Regions 0 and 2), a risk of loss of 0.8% (which is still lower than the project's maximum threshold of 2.5%, but stil a higher risk of loss than Regions 0 and 2), and an mean confidence interval of: approximately 
$757,716.90-$9,120,997.57.

### Conclusion:
In conclusion, it appears that the result from the predicted data and the bootstrapped data varied significantly.  According to the predicted data, Region 2 had the highest estimated profit of $33217543.96 and according to the boostrapped data, Region 1 had the highest average estimated profit (according to the avergae mean of the bootrapped samples) of: $4554662.44.